tools/unittestharness.py

```diff
@@ -1,120 +1,205 @@
+    testsResults = {'Pass': Counter(), 'Warn': Counter(), 'Fail': Counter()}
+    errorList, skipList = [], []
+
+    for test in tests:
+        print(f'\t{test.name}: Running')
+        testProc = subprocess.run(
+            './' + test.name, shell=True, cwd=test.parent, encoding='utf-8',
+            stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+
+        if re.search(r'^\s*error[:\s]', testProc.stderr,
+                     re.IGNORECASE | re.MULTILINE):
+            print(f'ERROR: {test.name} failed')
+            errorList.append(test.name)
+            continue
+        elif ': not found' in testProc.stderr:
+            print(f'WARNING: {test.name} not found')
+            skipList.append(test.name)
+            continue
+
+        for testOutput in testProc.stdout.splitlines():
+            try:
+                testsResults[testOutput[:4]][test.name] += 1
+            except KeyError:
+                print(testOutput.lstrip())
+                continue
+
+            print(f'\t\t{testOutput}')
+
+    printResults(testsResults, errorList, skipList)
+
+
+def unitTestHarness(tests, xml_outfile):
+    xmlParser = TestSuite('UnitTestHarness')
+    testsResults = {'Pass': Counter(), 'Warn': Counter(), 'Fail': Counter()}
+    errorList, skipList = [], []
+
+    for test in tests:
+        print(f'\t{test.name}: Running')
+        begin = process_time()
+        testProc = subprocess.run(
+            './' + test.name, shell=True, cwd=test.parent, encoding='utf-8',
+            stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+        wallTime = process_time() - begin
+
+        testId = f'unittest.{test.resolve().parts[-3]}.{test.name}'
+
+        if re.search(r'^\s*error[:\s]', testProc.stderr,
+                     re.IGNORECASE | re.MULTILINE):
+            print(f'ERROR: {test.name} failed')
+            errorList.append(test.name)
+            xmlEntry = TestCase('', classname=testId,
+                                elapsed_sec=wallTime, status='Error')
+            xmlEntry.add_error_info(message=testProc.stderr)
+            xmlParser.test_cases.append(xmlEntry)
+            continue
+        elif ': not found' in testProc.stderr:
+            print(f'WARNING: {test.name} not found')
+            skipList.append(test.name)
+            xmlEntry = TestCase('', classname=test.name,
+                                elapsed_sec=wallTime, status='Skipped')
+            xmlEntry.add_skipped_info(message='Not found')
+            xmlParser.test_cases.append(xmlEntry)
+            continue
+
+        otherOut = ''
+        for testOutput in testProc.stdout.splitlines():
+            try:
+                testsResults[testOutput[:4]][test.name] += 1
+            except KeyError:
+                print(testOutput.lstrip())
+                otherOut += testOutput.lstrip() + '\n'
+                continue
+
+            print(f'\t\t{testOutput}')
+
+            try:
+                testMsg = re.search(r'(?<=\[).+(?=\])', testOutput).group()
```

Comment:
Can you add a comment explaining what it searches for?

Reply:
Added.